ddtrace/internal/_rand.pyx

```diff
@@ -18,38 +18,75 @@ That's 10,000,000 seconds until we hit the period. That's 115 days of
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 
 
 Python 3.7:
-Name (time in ns)                       Min                 Max                Mean             StdDev              Median                IQR            Outliers  OPS (Mops/s)            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-rand64bits                         167.5956 (1.0)      211.3155 (1.0)      190.2803 (1.0)       9.5815 (1.0)      187.7187 (1.0)      11.4513 (1.0)          15;1        5.2554 (1.0)          52      100000
-random.randbits                    222.7103 (1.33)     367.4459 (1.74)     250.2699 (1.32)     26.5930 (2.78)     242.1607 (1.29)     26.4550 (2.31)          6;1        3.9957 (0.76)         36      100000
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+Name (time in ns)                     Min                 Max                Mean             StdDev              Median                IQR            Outliers  OPS (Mops/s)            Rounds  Iterations
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+rand64bits                        57.5356 (1.0)      151.9838 (1.0)       79.8573 (1.0)      20.5218 (1.0)       73.8342 (1.0)      24.0444 (1.0)         24;10       12.5223 (1.0)         147      100000
+random.getrandbits               111.7341 (1.94)     224.2096 (1.48)     137.4291 (1.72)     24.7072 (1.20)     126.5603 (1.71)     24.1269 (1.00)         15;6        7.2765 (0.58)         82      100000
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 """
 from libc.stdint cimport uint64_t
+import os
 
 from ddtrace import compat
 
 
-cdef uint64_t x = compat.getrandbits(64) ^ 4101842887655102017
+cdef uint64_t state
+cdef object pid = None
+
+
+cpdef _getstate():
+    return state
+
+
+cpdef seed():
+    global state
+    state = <uint64_t>compat.getrandbits(64) ^ <uint64_t>4101842887655102017
+
+
+cpdef rand64bits(check_pid=True):
+    if check_pid:
+        global pid
+        _pid = os.getpid()
+
+        if _pid != pid:
+            seed()
+
+        pid = _pid
```

Comment:
This is not thread-safe. Though I don't think it's a big issue, worst case you `seed()` twice.

Reply:
Good point. Yeah I think we're okay. Reseeding is no problem at all.